torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp

```diff
@@ -17,171 +18,881 @@ namespace fuser {
+        }
+        break;
+      }
+
+      if (inner_pos != pos_after_outer) {
+        // Nothing to the left of inner could be continuous.
+        //
+        // Clear left of inner
+        clear_left_of(find_inner_it);
+      }
+
+      if (hasPartialExtent(split->inner())) {
+        // Nothing to the left of inner can map, clear left of inner
+        clear_left_of(find_inner_it);
+      }
+
+      if (find_outer_it != ids.end() && find_inner_it != ids.end()) {
+        // Both dimensions map, inner dimension maps fully. For more context see
+        // the comment:
+        //  projectIdToRFactor
+        //    if (find_outer_it != ids.end() && find_inner_it != ids.end() &&
+        //      !hasPartialExtent(merge->inner())) {
+        //      Comment here.
+
+        // Removed outer, so inner shifts left
+        ids[inner_pos] = split->in();
+        ids.erase(find_outer_it);
+        auto in_pos = inner_pos - 1;
+        if (hasPartialExtent(split->outer())) {
+          // Outer dimension has a partial map, but inner dimension is full,
+          // split->in should be partially mapped.
+          partial_projected_extent_[split->in()] =
+              SimplifyingIrBuilder::mulExpr(
+                  getMaybePartialMappedExtent(split->outer()),
+                  split->inner()->extent());
+          // Clear to the left of split in since it partially maps
+          if (inner_pos > 0) {
+            ids.erase(ids.begin(), ids.begin() + in_pos);
+          }
+        }
+      } else {
+        // Only inner matches, mark a partial match
+        ids[inner_pos] = split->in();
+        // Mark the partial match, inner could already be a partial match
+        partial_projected_extent_[split->in()] =
+            getMaybePartialMappedExtent(split->inner());
+        // Clear to the left of split in since it partially maps
+        ids.erase(ids.begin(), ids.begin() + inner_pos);
+      }
+    } else if (const Merge* merge = dynamic_cast<const Merge*>(expr)) {
+      auto find_out_it = std::find(ids.begin(), ids.end(), merge->out());
+      if (find_out_it == ids.end()) {
+        continue;
+      }
+
+      auto out_pos = std::distance(ids.begin(), find_out_it);
+
+      if (!hasPartialExtent(merge->out())) {
+        // No partial map in output, so simply map through to inputs.
+        ids[out_pos] = merge->outer();
+        ids.insert(ids.begin() + out_pos + 1, merge->inner());
+      } else {
+        // This could be done completely symbolically by adding more expressions
+        // into the graph, for example the inner patial extent could be:
+        //   auto inner_partial_extent =
+        //   SimplifyingIrBuilder::minExpr(getMaybePartialMappedExtent(merge->out()),
+        //   merge->inner()->extent());
+        // and outer partial extent:
+        //   auto outer_partial_extent =
+        //   SimplifyingIrBuilder::divExpr(getMaybePartialMappedExtent(merge->out()),
+        //   merge->inner()->extent())
```

Comment:
I don't think this can be completely symbolicly. Depending on the evaluated value of the `outer_partial_extent`, we might want to just remove the outer id (for example, if `outer_partial_extent` is 0 or 1), also, although `merge->out()` is guaranteed to be divisible by `merge->inner()->extent()`, I don't think `getMaybePartialMappedExtent(merge->out())` has the same good property. Depending on the runtime value whether it is or not divisible or not, we might need to make different decisions on whether to keep the outer/inner ID.

Reply:
I'm not following you. I think this could be done all symbolically but seems like not worth the effort right now. I agree we could want to keep or drop different dimensions, but we could track all of that in conditionals and evaluate them at runtime like anything else. You're right we may or may not want to remove that outer id depending on the sizes, but just throw if statements in and evaluate those. I'm missing why we can't do this.

If outer or inner is partial we just remove everything to the left, so even if we keep outer but it's partially mapped, I don't know I see the problem as long as we evaluate if it's partial or not at all mapped. We can only keep one of these dimensions on the left most side of the mapped dims. I believe all we'd be debating is if outer partially maps, or if inner partially maps.

Reply:
> I don't think getMaybePartialMappedExtent(merge->out()) has the same good property. Depending on the runtime value whether it is or not divisible or not, we might need to make different decisions on whether to keep the outer/inner ID.

I think with only processing divisible splits this would be the case (will push code for this). Effectively what we need to say for these passes is that we can only expand vectorization over dimensions with divisible splits. Which I think makes sense on our conversation earlier this week. So we are effectively always maintaining divisibility.

Reply:
Still thinking this through, interesting example:
```
            // [3*5, 7*11] split->(0, 5)
            // [3, 5, 7*11] merge->(2)
            // [3, 5*7*11] split(1, 5)
            // [3, 7*11, 5]
            // If we start by mapping the inner most dimension of the last
            // domain. we have:
            // [3, 7*11, 5] 5 full maps
            // [3, 5*7*11] 5 partial maps
            // Now we would try to propagate
            // [3, 5*7*11] -> [3, 5, 7*11]
            // Problem here is we would partial map 5 to 7*11 but it is not
            // divisible by 5, so we can't vectorize based on that value.
```